File: websocket-api/javascript/create-order.js
const dotenv = require('dotenv')
const WebSocket = require('ws')
const crypto = require('crypto')

dotenv.config()

if (!process.env.API_KEY || !process.env.SECRET_KEY) {
  throw new Error('Required environment variables (API_KEY, SECRET_KEY) are missing.')
}

const apiKey = process.env.API_KEY
const secretKey = process.env.SECRET_KEY

// WebSocket endpoint
const WS_URL = 'wss://ws-api.ripio.com'

// Generate signature for WebSocket API
function generateSignature(timestamp, bodyParams) {
  const message = timestamp.toString() + JSON.stringify(bodyParams)
  return crypto.createHmac('sha256', secretKey).update(message).digest('base64')
}

// Create WebSocket connection
const ws = new WebSocket(WS_URL)

ws.on('open', () => {
  console.log('Connected to WebSocket API')

  // Prepare order parameters (business params only for signature)
  // Using invalid pair ABC_DEF for safety
  const timestamp = Date.now()
  const bodyParams = {
    pair: 'ABC_DEF',
    side: 'buy',
    type: 'limit',
    amount: 0.01,
    price: 300000
  }

  // Generate signature
  const signature = generateSignature(timestamp, bodyParams)

  // Create complete request with authentication
  const request = {
    id: 'req-create-001',
    method: 'order.create',
    params: {
      ...bodyParams,
      apiToken: apiKey,
      timestamp: timestamp,
      signature: signature
    }
  }

  console.log('\nSending create order request:', JSON.stringify(request, null, 2))
  ws.send(JSON.stringify(request))
})

ws.on('message', (data) => {
  const response = JSON.parse(data.toString())
  console.log('\nReceived response:', JSON.stringify(response, null, 2))
  
  // Close connection after receiving response
  setTimeout(() => {
    ws.close()
  }, 1000)
})

ws.on('error', (error) => {
  console.error('WebSocket error:', error)
})

ws.on('close', () => {
  console.log('\nWebSocket connection closed')
})

File: websocket-api/python/create_order.py
import asyncio
import json
import hmac
import hashlib
import base64
import time
import os
import websockets
from dotenv import load_dotenv

load_dotenv()

if os.getenv('API_KEY') is None or os.getenv('SECRET_KEY') is None:
    raise ValueError("Required environment variables (API_KEY, SECRET_KEY) are missing.")

api_key = os.getenv('API_KEY')
secret_key = os.getenv('SECRET_KEY')

# WebSocket endpoint
WS_URL = 'wss://ws-api.ripio.com'

def generate_signature(timestamp, body_params):
    """Generate HMAC SHA256 signature for WebSocket API"""
    message = str(timestamp) + json.dumps(body_params, separators=(',', ':'))
    signature = hmac.new(
        secret_key.encode(),
        message.encode(),
        hashlib.sha256
    ).digest()
    return base64.b64encode(signature).decode()

async def create_order():
    async with websockets.connect(WS_URL) as websocket:
        print('Connected to WebSocket API')

        # Prepare order parameters (business params only for signature)
        # Using invalid pair ABC_DEF for safety
        timestamp = int(time.time() * 1000)
        body_params = {
            'pair': 'ABC_DEF',
            'side': 'buy',
            'type': 'limit',
            'amount': 0.01,
            'price': 300000
        }

        # Generate signature
        signature = generate_signature(timestamp, body_params)

        # Create complete request with authentication
        request = {
            'id': 'req-create-001',
            'method': 'order.create',
            'params': {
                **body_params,
                'apiToken': api_key,
                'timestamp': timestamp,
                'signature': signature
            }
        }

        print('\nSending create order request:', json.dumps(request, indent=2))
        await websocket.send(json.dumps(request))

        # Receive response
        response = await websocket.recv()
        response_data = json.loads(response)
        print('\nReceived response:', json.dumps(response_data, indent=2))

if __name__ == '__main__':
    asyncio.run(create_order())
